server/src/main/java/org/elasticsearch/action/ActionModule.java

```diff
@@ -477,24 +478,26 @@ public ActionModule(
                 new RestHeaderDefinition(Task.X_ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER, false)
             )
         ).collect(Collectors.toSet());
-        UnaryOperator<RestHandler> restWrapper = null;
+        UnaryOperator<RestHandler> restInterceptor = null;
         for (ActionPlugin plugin : actionPlugins) {
-            UnaryOperator<RestHandler> newRestWrapper = plugin.getRestHandlerWrapper(threadPool.getThreadContext());
-            if (newRestWrapper != null) {
-                logger.debug("Using REST wrapper from plugin " + plugin.getClass().getName());
-                if (plugin.getClass().getCanonicalName() == null
-                    || plugin.getClass().getCanonicalName().startsWith("org.elasticsearch.xpack") == false) {
-                    throw new IllegalArgumentException(
-                        "The "
-                            + plugin.getClass().getName()
-                            + " plugin tried to install a custom REST "
-                            + "wrapper. This functionality is not available anymore."
-                    );
+            if (plugin instanceof RestInterceptor riplugin) {
```

Comment:
With this change, there is no inherent relationship between `ActionPlugin` and `RestInterceptor`. So how can we be sure this `instanceof` check is only needed inside a loop for ActionPlugins? It is probably not an issue at all and mainly due to my lack of understanding of future modules change. I appreicate if you could help explaining it. Thanks!

Reply:
Yeah, this is a good point. I think it would be better to make RestInterceptor extend ActionPlugin (since it is really a specialised action plugin, rather an a standalone entity )